server/src/main/java/org/apache/iotdb/db/metadata/MTree.java

```diff
@@ -781,20 +777,56 @@ private void findStorageGroup(MNode node, String path, HashSet<String> res) {
    *
    * @return a list contains all distinct devices
    */
-  Set<String> getAllDevices() {
-    return new HashSet<>(getNodesList(3));
+  Set<String> getAllDevices() throws SQLException {
+    List<String> res = new ArrayList<>();
+    MNode node;
+    if ((node = getRoot()) != null) {
+      findDevices(node, "root", res);
+    }
+    return new LinkedHashSet<>(res);
+  }
+
+  private void findDevices(MNode node, String path, List<String> res) {
+    if (node == null) {
+      return;
+    }
+    if (node.isLeaf()) {
+      res.add(path);
+      return;
+    }
+    if (node.hasChildren()) {
+      for (MNode child : node.getChildren().values()) {
+        if (child.isLeaf()) {
+          findDevices(child, path, res);
+        } else {
+          findDevices(child, path + "." + child.toString(), res);
+        }
+      }
+    }
   }
 
   /**
    * Get all nodes at the given level in current Metadata Tree.
    *
    * @return a list contains all nodes at the given level
    */
-  List<String> getNodesList(int nodeLevel) {
+  List<String> getNodesList(String schemaPattern, int nodeLevel) throws SQLException {
```

Comment:
Is this function only for the count? If so, why not return an int?

Reply:
No, it is not only for the count. The function "COUNT NODE ... GROUP BY ..."  has also been shared with this method, as it needs paths (path string) in this list so that it could show the node number within the specified path under the specified level.

Reply:
ok